particles/PipeApps/source/RecentAddresses.js

```diff
@@ -15,25 +15,17 @@ defineParticle(({DomParticle, html, log}) => {
     get template() {
       return html`<div slotid="assistance"></div>`;
     }
-    update({recentAddresses}, state) {
-      if (recentAddresses) {
-        const address = this.query(recentAddresses);
+    update({recentEntities}, state) {
+      if (recentEntities) {
+        const address = this.query(recentEntities);
         this.updateVariable('address', {address});
       }
     }
-    query(addresses) {
-      //const sorted = addresses.sort((a, b) => (b.rawData.timestamp || 0) - (a.rawData.timestamp || 0));
-      // const sliced = addresses.slice(0, 3);
-      // return JSON.stringify(sliced.map(e => e.rawData));
-      if (addresses.length) {
-        const address = addresses[addresses.length-1];
-        const result = {
-          type: 'address',
-          name: address.rawData.address,
-          source: 'com.weaseldev.fortunecookies'
-        };
-        return JSON.stringify(result);
-      }
+    query(entities) {
+      const addresses = entities.filter(entity => entity.rawData.type === 'address');
+      const sorted = addresses.sort((a, b) => (b.rawData.timestamp || 0) - (a.rawData.timestamp || 0));
+      const result = sorted[0] || Object;
+      return JSON.stringify(result.rawData);
```

Comment:
If we end up with just `Object` won't this line fail dereferencing `rawData`.

Reply:
No, trying to deref a property off of a non-object (like null) throws an error, but dereferencing an undefined property on any object just returns undefined. `Object` makes a convenient null-object in this scenario (and it's slightly faster than `{}`);

Reply:
Just to verify
![image](https://user-images.githubusercontent.com/98045/53596009-98990400-3b53-11e9-9f87-46220851cffc.png)